tmt/base.py

```diff
@@ -2453,11 +2476,171 @@ def export(self):
         return data
 
 
-class Link(tmt.utils.SerializableContainer):
-    """ Core attribute link parsing """
+@dataclasses.dataclass
+class LinkNeedle:
+    """
+    A container to use for searching links.
+
+    ``relation`` and ``target`` fields hold regular expressions that
+    are to be searched for in the corresponding fields of :py:class:`Link`
+    instances.
+    """
+
+    relation: str = r'.*'
+    target: str = r'.*'
+
+    @classmethod
+    def from_spec(cls, value: str) -> 'LinkNeedle':
+        """
+        Create a ``LinkNeedle`` instance from its specification.
+
+        Specification is described in [1], this constructor takes care
+        of parsing it into a corresponding ``LinkNeedle`` instance.
+
+        [1] https://tmt.readthedocs.io/en/stable/spec/plans.html#fmf
+        """
+
+        parts = value.split(':', maxsplit=1)
+
+        if len(parts) == 1:
+            return LinkNeedle(target=parts[0])
+
+        return LinkNeedle(relation=parts[0], target=parts[1])
+
+    def __str__(self) -> str:
+        return f'{self.relation}:{self.target}'
+
+    def matches(self, link: 'Link') -> bool:
+        """ Find out whether a given link matches this needle """
+
+        if not re.search(self.relation, link.relation):
+            return False
+
+        if isinstance(link.target, str):
+            return re.search(self.target, link.target) is not None
+
+        # TODO: how to match fmf id?
+        return False
```

Comment:
This I did not catch in the original implementation, I'm not sure how matching `link: "relation:target"` against links with fmf id as their target worked.

```python
        for candidate in self.link.get():
            candidate_relation = get_relation(candidate)
            candidate_target = candidate[candidate_relation]
            if (re.search(relation, candidate_relation)
                    and re.search(target, candidate_target)):
                return True
        return False
```

Links with fmf id as their target should fail when facing `re.search(target, candidate_target)` call as `candidate_target` would be a dictionary. Maybe there's some implicit conversion I'm missing, in any case, I'd propose to make it explicit. I left this bit empty in my patch, and I'd like to hear from tmt folks how the `relates:some-test-[a-z]+` link search should test fmf ids. I can imagine checking all non-`None` fields of the id, for example, or some form of `str(fmf_id)` and checking that (which would be the same thing, just a different implementation...)

Reply:
Hm, this is a tricky one! So before we just completely ignored all fmf identifiers during the search:

https://github.com/teemtee/tmt/blob/a2b8ca7a398fa5137117fab8eb34707db520af0d/tmt/base.py#L1549-L1551

In order to allow exact search we would have to add support for checking individual fmf id fields, that means perhaps something like this:

   tmt test ls --link verifies:url=https://example.org;name=/some/name;ref=branch

I'd suggest to not block this PR on implementing this (needs discussion, not high priority, not regression). Perhaps, we could go with just a simple implementation for now: If the fmf id key is not specified, go with matching against `name` only. Would that make sense?

Reply:
So, if I understand it correctly, this would be good enough for now:

```python
    def matches(self, link: 'Link') -> bool:
        """ Find out whether a given link matches this needle """

        # Rule out the simple case, mismatching relation.
        if not re.search(self.relation, link.relation):
            return False

        # If the target is a string, the test is trivial.
        if isinstance(link.target, str):
            return re.search(self.target, link.target) is not None

        # If the target is an fmf id, the current basic implementation will
        # check just the `name` key, if it's defined. More fields may come
        # later, pending support for more sophisticated parsing of link
        # needle on a command line.
        if link.target.name:
            return re.search(self.target, link.target.name) is not None

        return False
```

Reply:
Yes, exactly.

Reply:
Ack, fixed.